src/opuntia.erl

```diff
@@ -1,53 +1,106 @@
 %%
 %% This function takes the current shaper state, and the number of tokens that have been consumed,
 %% and returns a tuple containing the new shaper state, and a possibly non-zero number of
 %% unit times to wait if more tokens that the shaper allows were consumed.
--spec update(shaper(), tokens()) -> {shaper(), rate()}.
-update(none, _Size) ->
+-spec update(shaper(), tokens()) -> {shaper(), delay()}.
+update(none, _TokensNowUsed) ->
     {none, 0};
-update(#token_bucket{rate = MaxRatePerMs,
-                     available_tokens = LastAvailableTokens,
-                     last_update = LastUpdate} = Shaper, TokensNowUsed) ->
-    Now = erlang:monotonic_time(millisecond),
-    % How much we might have recovered since last time
+update(#token_bucket_shaper{shape = {MaximumTokens, Rate, TimeUnit},
+                            available_tokens = LastAvailableTokens,
+                            last_update = LastUpdate} = Shaper, TokensNowUsed) ->
+
+    %% Time since last shape update
+    Now = erlang:monotonic_time(TimeUnit),
     TimeSinceLastUpdate = Now - LastUpdate,
-    PossibleTokenGrowth = round(MaxRatePerMs * TimeSinceLastUpdate),
-    % Available plus recovered cannot grow higher than the actual rate limit
-    ExactlyAvailableNow = min(MaxRatePerMs, LastAvailableTokens + PossibleTokenGrowth),
-    % Now check how many tokens are available by substracting how many where used,
-    % and how many where overused
+
+    %% How much we might have recovered since last time
+    AvailableAtGrowthRate = Rate * TimeSinceLastUpdate,
+    UnboundedTokenGrowth = LastAvailableTokens + AvailableAtGrowthRate,
+
+    %% Real recovery cannot grow higher than the actual rate in the window frame
+    ExactlyAvailableNow = min(MaximumTokens, UnboundedTokenGrowth),
+
+    %% How many are available after using TokensNowUsed can't be smaller than zero
     TokensAvailable = max(0, ExactlyAvailableNow - TokensNowUsed),
+
+    %% How many tokens I overused might be zero if I didn't overused any
     TokensOverused = max(0, TokensNowUsed - ExactlyAvailableNow),
-    MaybeDelay = TokensOverused / MaxRatePerMs,
-    RoundedDelay = floor(MaybeDelay) + 1,
-    NewShaper = Shaper#token_bucket{available_tokens = TokensAvailable,
-                                    last_update = Now + MaybeDelay},
+
+    %% And then MaybeDelay will be zero if TokensOverused was zero
+    MaybeDelay = TokensOverused / Rate,
+
+    %% We penalise rounding up, the most important contract is that rate will never exceed that
+    %% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
+    %% don't promise at what time in the future they would arrive, nor we promise any upper bound
+    %% to the limits of the shaper delay.
+    RoundedDelay = ceil(MaybeDelay),
+
+    NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
+                                           last_update = Now + RoundedDelay + 1},
```

Comment:
Oh, this looks incorrect instantly. You need to store the accumulated fraction, because otherwise it would block you right away. Let's see:

```
8> lists:foldl(fun(N, ShIn) -> {ShOut, 0} = opuntia:update(ShIn, N), ShOut end, opuntia:new({100000,10000,second}), [1,1]).
** exception error: no match of right hand side value {#token_bucket_shaper{shape = {100000,10000,second},
                                                                            available_tokens = 0,last_update = -576460643},
                                                       1}
```

My hypothesis was instantly that for `millisecond` you would need 1000 such requests:

```
31> lists:foldl(fun(N, ShIn) -> {ShOut, 0} = opuntia:update(ShIn, N), ShOut end, opuntia:new({100000,10000,millisecond}), [1,1]).
** exception error: no match of right hand side value {#token_bucket_shaper{shape = {100000,10000,millisecond},
                                                                            available_tokens = 0,last_update = -576460202884},
                                                       1}
```

Only `microsecond` would save situation:

```
21> lists:foldl(fun(N, ShIn) -> {ShOut, 0} = opuntia:update(ShIn, N), ShOut end, opuntia:new({100000,10000,microsecond}), lists:duplicate(1000,1)).
#token_bucket_shaper{shape = {100000,10000,microsecond},
                     available_tokens = 89991,last_update = -576460237904693}
```

Reply:
A draft of what I meant by fractions is below. It passed my quick test in the console, but of course unit tests would show if it works:

```
--- a/src/opuntia.erl
+++ b/src/opuntia.erl
@@ -60,8 +60,9 @@ new(0) ->
 new({MaximumTokens, Rate, TimeUnit})
   when ?NON_NEG_INT(MaximumTokens), ?NON_NEG_INT(Rate), ?TU(TimeUnit) ->
     #token_bucket_shaper{shape = {MaximumTokens, Rate, TimeUnit},
-                         available_tokens = 0,
-                         last_update = erlang:monotonic_time(TimeUnit)}.
+                         available_tokens = Rate,
+                         last_update = erlang:monotonic_time(TimeUnit),
+                         frac = 0.0}.

 %% @doc Update shaper and return possible waiting time.
 %%
@@ -73,11 +74,12 @@ update(none, _TokensNowUsed) ->
     {none, 0};
 update(#token_bucket_shaper{shape = {MaximumTokens, Rate, TimeUnit},
                             available_tokens = LastAvailableTokens,
-                            last_update = LastUpdate} = Shaper, TokensNowUsed) ->
+                            last_update = LastUpdate,
+                            frac = Frac} = Shaper, TokensNowUsed) ->

     %% Time since last shape update
     Now = erlang:monotonic_time(TimeUnit),
-    TimeSinceLastUpdate = Now - LastUpdate,
+    TimeSinceLastUpdate = (Now - LastUpdate) + Frac,

     %% How much we might have recovered since last time
     AvailableAtGrowthRate = Rate * TimeSinceLastUpdate,
@@ -102,5 +104,6 @@ update(#token_bucket_shaper{shape = {MaximumTokens, Rate, TimeUnit},
     RoundedDelay = ceil(MaybeDelay),

     NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
-                                           last_update = Now + RoundedDelay + 1},
+                                           last_update = Now + RoundedDelay,
+                                           frac = RoundedDelay - MaybeDelay},
     {NewShaper, RoundedDelay}.
diff --git a/src/opuntia.hrl b/src/opuntia.hrl
index aff644f..3ec40f1 100644
--- a/src/opuntia.hrl
+++ b/src/opuntia.hrl
@@ -2,9 +2,10 @@
 -define(OPUNTIA, true).

 -record(token_bucket_shaper, {
+          frac :: float(),
           shape :: opuntia:shape(),
           available_tokens :: opuntia:tokens(),
-          last_update :: number()
+          last_update :: integer()
 }).
```